### ISSUE TITLE
Run hanging in while loop

### ISSUE THREAD

**Issue by akijarl** [mossmatters/HybPiper#33]
Hello,

I'm hoping to run the HybPiper pipeline on my institution's cluster. After testing that all dependencies were available and reads_first.py was running normally on a sand-box partition of the cluster, I submitted a job on the main partition using the while loop specified below:

```
while read name;
do ../reads_first.py -b GenesForCapture.fasta -r $name*.fastq --prefix $name --bwa
done < Sample_names.txt
```

The first sample finished running in about 2 hours without a problem, however the second sample appears to still be running after some 30 hours, despite all gene folders for the second sample having a /sequences/FNA/ and /sequences/FAA/ output.

The most recent output is below:

> ^META: 479s Left: 1 AVG: 479.08s  local:1/277/100%/479.1s sh: /dev/tty: No such device or address
> ^META: 479s Left: 1 AVG: 479.09s  local:1/277/100%/479.1s sh: /dev/tty: No such device or address
> ^META: 479s Left: 1 AVG: 479.09s  local:1/277/100%/479.1s sh: /dev/tty: No such device or address
> ^META: 479s Left: 1 AVG: 479.09s  local:1/277/100%/479.1s sh: /dev/tty: No such device or address
> ^META: 479s Left: 1 AVG: 479.10s  local:1/277/100%/479.1s sh: /dev/tty: No such device or address

The input files are paired end illumina sequence reads that were cleaned using the Illumiprocessor wrapper of Trinnotate.

Any help in understanding the hangup would be much appreciated.

**Comment by mossmatters** [comment]
The `/dev/tty` error suggests that something is happening with output generated when the job is submitted to the compute node.

I have seen some issues running HybPiper on a cluster, because of the high Input-Output demands. Some users have been able to solve this by utilizing local "scratch space" on the compute node for running HybPiper, and then copying the files back to the main/head/storage node at the end of the job.  

If you are already utilizing scratch space, then it may be filling up from the results of the first run. You can run `cleanup.py` after running `reads_first.py` to greatly reduce the size of the directory generated by HybPiper. 

Have you tried running the test dataset via the compute nodes, rather than in the sandbox?

Let me know if this helps -- it is difficult to keep track of how well HybPiper performs on HPCC systems, so I would appreciate the feedback!

**Comment by mossmatters** [comment]
Rebecca, have you tried using the `--timeout` command? With this you can have it automatically kill one of the Spades jobs if it takes more than, say, 4000% longer than average (`--timeout 4000`). 

You should not run `cleanup.py` while the main script is still running already running on the same sample. I was suggesting that the job should run using storage (hard drive) space that is unique to the compute node, to avoid errors in reading in/out files. On some clusters this is located in `/scratch` but your cluster may be different. If my reads are located in my home directory `~/reads', then a job submission script I might run would be something like:

```
cd /scratch
python HybPiper/reads_first.py -r ~/reads/1.fq ~/reads/2.fq -b ~/targets.fasta --bwa --prefix name
python HybPiper/cleanup.py --prefix name
cp -r name ~/
```
You can then run that in a loop over all of your samples. I suggested running cleanup.py after one sample is complete in case the storage space on `/scratch` is limited.

**Comment by stubbsrl** [comment]
Awesome! Thanks for the answer Matt! This is great info so I will give this a shot and will pass this  on to my colleagues.

As an important follow up the reason this was of concern was because I was worried it would mess with the post processing scripts, but as of yesterday everything ran perfectly for pulling out all the genes and getting paralogs, stats, and the heat map. This is with the samples that were stalled and I therefore had to kill the jobs.  So, it is likely this issue is more of a nuisance than an actual problem.

**Comment by mossmatters** [comment]
Better handling of timeouts, including issues related to low-complexity DNA in the target file, will be addressed in HybPiper 2.0
